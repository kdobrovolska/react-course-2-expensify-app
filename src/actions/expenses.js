import uuid from 'uuid';

export const AddExpense=({description='',note='',amount=1,createdAt=0}={})=>(
    {
        type:'ADD_EXPENCE',
        expense:{id:uuid(),description,note,amount,createdAt}
    }
);
export const RemoveExpense=(id)=>(
    {
        type:'REMOVE_EXPENCE',
        expenseID:id
    }
);
export const EditExpense=(id, updates={})=>(
    {
        type:'EDIT_EXPENCE',
        expenseID:id,
        expense: updates
    }
);