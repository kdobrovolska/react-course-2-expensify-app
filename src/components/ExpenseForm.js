import React from 'react';
import moment from 'moment';
const now =moment();
console.log(now);
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css'; //go to app.js

export default class ExpenseForm extends React.Component
{
    
    constructor(props)
    {
        
      //  alert('conzstructor');
        super(props);
        this.state={description:'',
            note:'', 
            amount:'',
            createdAt:moment(),
            calendarFocused:false,
            error:''
            };
        
        if(!!props.expense)
        {
          this.state =  {...this.state,...props.expense,createdAt:moment(props.expense.CreateAt)};
           
        }
    }
    
    DescriptionChanged=(e)=>
    {
        
        const description=e.target.value;
      //  alert(description);
        this.setState(()=>({description}));
    }
    NoteChanged=(e)=>
    {
        e.persist();
         this.setState(()=>({note:e.target.value}));
    }
    AmountChanged=(e)=>
    {
        
        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    }
    onDateChange=(createdAt) =>{
        if(createdAt)
            this.setState(()=>({createdAt}));
    }
    onFocusChange1=({focused})=>
    {
       this.setState(()=>({calendarFocused:focused}));
    }
    onSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount)
        {
            this.setState(()=>({error:'Please provide description and amount'}));
       
        }else{
            this.setState(()=>({error:''}));
            this.props.submit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                note:this.state.note,
                createdAt:this.state.createdAt.valueOf()
            });
        }
    };
    render()
    {
        let obj=this.state;
     /*   if(!!this.props.expense)
        {
           obj=  {...obj,...this.props.expense,createdAt:moment(this.props.expense.CreateAt)};
        }*/

     //   alert('render');
      return <form className='form' onSubmit={ this.onSubmit}>
       {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input type='text' 
            className='text-input' 
            placeholder='Description' 
            autoFocus
            value={obj.description}
            onChange={this.DescriptionChanged}
            />
        <label id='errDescription' className='err'></label>
        <input type='text' 
            className='text-input' 
            placeholder='Amount' 
            value={obj.amount}
            onChange={this.AmountChanged}
            />
            
         <SingleDatePicker 
            date={obj.createdAt}
            onDateChange={this.onDateChange}
            focused ={this.state.calendarFocused}
            onFocusChange ={this.onFocusChange1}
            numberOfMonths={1}
            isOutsideRange={()=>{ return false;}}
            />
        <p></p>
        <textarea placeholder='Add note' 
            className='textarea'
            value={obj.note}
            onChange={this.NoteChanged}
        ></textarea>
        <div>
            <button className='button1'>Save Expense</button>
        </div>
  
        </form> ;
    }
}