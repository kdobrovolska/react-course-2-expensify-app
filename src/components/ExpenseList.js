import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem.js'
import VisibleExpenses from '../selectors/expenses.js'

export const ExpenseList=(props)=>
(
    <div>
    <h1> Expence List  </h1>
     {props.expenses.map((item,ident)=>{
         
         return <ExpenseListItem key={ident} {...item}/>
    })}
    </div>
);
// first variant
//const ConnectedExpenseList=connect((st)=>{
  //  return {
  //      expenses:st.expenses
  //  };
//})(ExpenseList);
//export default ConnectedExpenseList;

const mapStateToProps=(state)=>
{
    return{
        expenses:VisibleExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);