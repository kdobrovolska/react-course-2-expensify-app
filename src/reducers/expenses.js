const expenseReducerDefaultState=[];

export default (state=expenseReducerDefaultState,action)=>{
    let ar;
    switch(action.type){
        case 'ADD_EXPENCE':
            return [...state,action.expense];
        case 'REMOVE_EXPENCE':
              return state.filter(({id})=>{return id!=action.expenseID});
        case 'EDIT_EXPENCE':
     //   alert('id='+action.expenseID);
           return state.map((item)=>{
               if(item.id===action.expenseID)
               {
                  // alert(action.expense.description);
                   return {...item,...action.expense, id:action.expenseID}
               }
               else
               {
                   return item;
               }
            })
        case 'SORT_BY_AMOUNT':
          ar=[...state];
          ar.sort((item1,item2)=>( item1.amount-item2.amount));
          return ar;
        case 'SORT_BY_DATE':
          ar=[...state];
          ar.sort((item1,item2)=>( item1.createdAt-item2.createdAt));
          return ar;
        default:
         return state;
    }
};