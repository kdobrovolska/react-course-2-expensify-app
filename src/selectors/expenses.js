import moment from 'moment'
export default (expenses,{startDate,endDate,text, sortBy})=>
{
  /*  const ar= [];
    expenses.map((item)=>{
        
        const isStartDate= (typeof startDate != 'number' ) || item.createdAt>=startDate;
        const isEndDate= (typeof endDate != 'number' ) || item.createdAt<=endDate;
        const isText = true; //(typeof endDate != 'string') || item.description.toLowerCase().indexOf(text.toLowerCase())>=0;
     //   alert(isStartDate+' , '+isEndDate+' , '+ isText );
      //  alert(startDate+' , '+endDate+' , '+ text +'=='+item.createdAt);
        if( isStartDate && isEndDate && isText)
        {
            ar.push(item);
        }
    });
   
    console.log('count='+expenses.length, ar.length);
   return ar;*/
   return expenses.filter((item)=>{
        const createdAtMoment=moment(item.createdAt);
        const isStartDate=  startDate? startDate.isSameOrBefore(createdAtMoment, 'day'):true;
        const isEndDate= endDate? endDate.isSameOrAfter(createdAtMoment, 'day'):true;
        const isText = (typeof text != 'string') || item.description.toLowerCase().includes(text.toLowerCase());
        return isStartDate && isEndDate && isText;
    }).sort((a,b)=>{
        if(sortBy.toLowerCase()==='date')
            return  (a.createdAt-b.createdAt);
        if(sortBy.toLowerCase()==='amount')
            return  (a.amount-b.amount);
  
    }); 
    
}