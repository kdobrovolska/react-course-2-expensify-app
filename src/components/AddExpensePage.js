import React from 'react';

import {connect} from 'react-redux';
import {AddExpense} from '../actions/expenses.js';
import ExpenseForm from './ExpenseForm';

/* const AddExpensePage = (props) => (
  <div>
    <h1> Add Form:</h1>
        <ExpenseForm 
           submit={(obj)=>  {
            console.log(obj);
            props.dispatch(AddExpense(obj));
            props.history.push('/');
        } } 
      />
  </div>
);
export default connect()(AddExpensePage); */


export class  AddExpensePage extends React.Component {
  onSubmit=(expense)=>{
      this.props.AddExpense(expense);
      this.props.history.push('/');
  }

  render()
  {  return   <div>
    <h1> Add Form:</h1>
        <ExpenseForm  submit={ this.onSubmit} />
    </div>;
  }
}
const MapDispatchToProps =(dispatch)=>{
 return { AddExpense:(expense)=>dispatch(AddExpense(expense))}
}
export default connect(undefined,MapDispatchToProps)(AddExpensePage);