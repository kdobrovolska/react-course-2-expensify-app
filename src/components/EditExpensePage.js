import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import ExpenseForm from './ExpenseForm';
import {EditExpense, RemoveExpense} from '../actions/expenses.js';

export class EditExpensePage extends React.Component{

   submit=(obj)=>  {
          this.props.EditExpense(this.props.expense.id,obj);
          this.props.history.push('/');
        }

   onClick=()=>{
          this.props.RemoveExpense(this.props.expense.id);
          this.props.history.push('/');
        }

    render()
    {
        return <div>
            <ExpenseForm expense={this.props.expense}
              submit={this.submit } />
          <button onClick={ this.onClick}>Remove </button>
          Privet
        </div>;
    }
};

const MapStoreToProps=(state,props)=>
{
     return { expense:state.expenses.filter(
       (item)=>(item.id===props.match.params.id))[0]
     }
}

 const MapDispatchToProps =(dispatch)=>{
  return { 
    EditExpense:(id,expense)=>dispatch(EditExpense(id,expense)),
    RemoveExpense: (id)=>dispatch(RemoveExpense(id))
   }
 } 

export default connect(MapStoreToProps, MapDispatchToProps)(EditExpensePage);
//export default connect(MapStoreToProps)(EditExpensePage);