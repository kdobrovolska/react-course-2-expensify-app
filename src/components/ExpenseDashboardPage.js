import React from 'react';
import ExpenseList from './ExpenseList.js'
import ExpenseListFilters from './ExpenseListFilters.js'

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList type='new'/>
  </div>
);
export default ExpenseDashboardPage;