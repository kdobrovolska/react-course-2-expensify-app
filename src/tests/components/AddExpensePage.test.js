import React from 'react'
import {shallow} from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage'
import {ExpenseForm} from '../../components/ExpenseForm'
import expenses from '../fixtures/expenses'
let AddExpense,history,wrapper;

beforeEach(()=>{
    AddExpense= jest.fn();
     history ={push: jest.fn()};
     wrapper =shallow(<AddExpensePage AddExpense={AddExpense} history={history}/>);
});

test('should render AddExenseForm correctly', ()=>{
    
    expect(wrapper).toMatchSnapshot();
});

test ('should test add data', ()=>{
    wrapper.find('ExpenseForm').prop('submit')(expenses[0]);
   expect(history.push).toHaveBeenLastCalledWith('/');
   expect(AddExpense).toHaveBeenLastCalledWith(expenses[0]);
});