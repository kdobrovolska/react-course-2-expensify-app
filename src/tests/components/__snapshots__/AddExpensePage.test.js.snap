// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExenseForm correctly 1`] = `
<div>
  <h1>
     Add Form:
  </h1>
  <ExpenseForm
    submit={[Function]}
  />
</div>
`;
